README
CHANGELOG:

6-Sept-2014
Thanks to Cambell Prince
- Added composer.json and dependency on Slim
- Improved error presentation for xdebug users.
- Changed expected headers to uppercase.
- Switch to use composer installed Slim
- Switch to use Slim installed via composer.
- Remove hard coded path 'api' and simplify includes.

17-Sept-2014
Thanks to Salman Sarwar
- Bug Fix in inventory.inc

14-July-2013:
- Added .htaccess so you can now use API URL's without index.php, examples: (Thanks to Christian Estrella)
	OLD: GET http://mysystem.com/api/index.php/locations/
	NEW: GET http://mysystem.com/api/locations/

- Added Pagination to GET methods, it used to return all entries, now is per page, under index.php it has define("RESULTS_PER_PAGE", 2); that defines how many entries you will get per page, if you dont establish a page on the request you will get the first page. (Thanks to Christian Estrella)
	OLD Request: GET http://mysystem.com/api/index.php/locations/
	OLD Response: ALL LOCATIONS
	
	NEW Request 1: GET http://mysystem.com/api/index.php/locations/
	NEW Response 1: First Page of Locations
	
	NEW Request 2: GET http://mysystem.com/api/index.php/locations/?page=5
	NEW Response 2: Page 5 of Locations

- Added Sales Transactions Methods for Quotes, Sales Orders, Deliveries, Invoices (GET, PUT, POST)
	NOTE: This changes hasn't been tested deeply, might have some bugs

14-June-2013:
- Added POST /locations/ To Add A Location Thanks to Richard Vinke


README:

# This is a very Simple Front Accounting API

I needed some basic integration functions to another software and decided to create this REST API and contribute with Front Accounting team.
Hope you find it usefull !!

== API Quick Start ==

1. Just copy the files into the modules directory under a folder called "api" or anything you want.
2. Edit the file util.php and change the $company, $username and $password variables so you can test. Use it at your own risk, to provide login from another software you need to send X-COMPANY, X-USER and X-PASSWORD headers in the request and the API will use those credentials, if they're wrong you will get a nice message telling "Bad Login"
3. Try to access the API, for example, try the Items Category List, type this on your explorer: http://YOUR_FA_URL/modules/api/category/ You will see a JSON with all you're items categories, if not check the util.php file.

== Methods ==

So far i have implemented Items, Inventory Movements, Locations, Items Categories, Tax Groups, Tax Types, Bank Accounts, GL Accounts, Customers, Suppliers and GL Account Types. Some of them have not been tested yet so be carefull.

== Contact ==

Any question about this you can always contact me: andres.amaya.diaz@gmail.com

CHANGELOG.md
# CHANGELOG

#### July 2018

- 03 Jul: Fixed #46 Trial Balance feature at endpoint /glquery/trialbalance/{startDate}/{endDate}
- 03 Jul: Fixed #47 Journal post without having a memo set on an item causes 500 server error
- 03 Jul: Complete #43, add journal support for non-company currencies and exchange rate.

#### June 2018

- 29 Jun: In Journal add support for get, post, put of memo.
- 29 Jun: Use ISO8601 for Journal dates.
- 28 Jun: Fix #41 Exchange rate get (all) doesn't return an ISO8601 date.

#### 26 June 2018 V2.4-1.7

- 26 Jun: Add full CRUD for exchange rates under a new /exchangerates endpoint.

#### 22 June 2018 v2.4-1.6

- 23 Jun: Change README.md API documentation to point to http://andresamayadiaz.github.io/FrontAccountingSimpleAPI/
- 22 Jun: Add full CRUD for Journal entry, update, and delete (void).
- 21 Jun: Add full CRUD for Bank Accounts under bankaccounts endpoint.
- 21 Jun: Fix formatting of code under src/ and test/ to conform to [PSR-2](https://www.php-fig.org/psr/psr-2/)
- 20 Jun: Add full CRUD for GLAccounts under glaccounts endpoint.
- 20 Jun: Add CRUD for Dimensions under dimensions endpoint.
- 20 Jun: Add swagger and spectacle to produce api documentation from annotated code.

#### 19 June 2018 v2.4-1.5

- Core VARLIB_PATH and VARLOG_PATH inclusions added [see this commit](https://github.com/FrontAccountingERP/FA/commit/4a37a28c49bf900dcc370fd3f21186cedcd632c9).
- TaxType: Added getById (Apmuthu 24 Nov 2017).
- Sales: Fix #32 Sales transactions Total not returned.
- Stock Adjust: Added unit test.
- Stock Adjust: Return now encoded as json msg.
- Stock Adjust: Missing argument $info fixed (Apmuthu 23 Apr 2018).
- Stock Adjust: add_stock_adjustment parameters fixed for FA 2.4 changes in API (Apmuthu, justapeddler 19 Apr 2018).
- Translated Spanish comments to English (Apmuthu 18 Nov 2017).

#### 27 November 2017

- Refactor tests to ensure consistency in POST and PUT api.
  Added tests/Crud_Base.php
- Add api_ensureAssociateArray to remove the numeric index elements that come from the Front Accounting functions.
- The category end point now follows the database schema more closely for property names.
- The customers end point now follows the database schema more closely for property names.

#### 23 November 2017

- Added support for requests sent using Content-Type: application/json
  The body is presumed to be in JSON format and converted appropriately.

#### 17 November 2017

- Updated API to support Front Accounting version 2.4.x
- Added PHP Unit tests
- Added Travis CI build

#### 6 September 2014
Thanks to Cambell Prince

- Added composer.json and dependency on Slim
- Improved error presentation for xdebug users.
- Changed expected headers to uppercase.
- Switch to use composer installed Slim
- Switch to use Slim installed via composer.
- Remove hard coded path 'api' and simplify includes.

#### 17 September 2014
Thanks to Salman Sarwar

- Bug Fix in inventory.inc

#### 14 July 2013:
- Added .htaccess so you can now use API URL's without index.php, examples:
  (Thanks to Christian Estrella)
    OLD: GET http://mysystem.com/api/index.php/locations/
    NEW: GET http://mysystem.com/api/locations/

- Added Pagination to GET methods, it used to return all entries, now is per page, under index.php it has define("RESULTS_PER_PAGE", 2); that defines how many entries you will get per page, if you dont establish a page on the request you will get the first page. (Thanks to Christian Estrella)
    OLD Request: GET http://mysystem.com/api/index.php/locations/
    OLD Response: ALL LOCATIONS
    
    NEW Request 1: GET http://mysystem.com/api/index.php/locations/
    NEW Response 1: First Page of Locations
    
    NEW Request 2: GET http://mysystem.com/api/index.php/locations/?page=5
    NEW Response 2: Page 5 of Locations

- Added Sales Transactions Methods for Quotes, Sales Orders, Deliveries, Invoices (GET, PUT, POST)
    NOTE: This changes hasn't been tested deeply, might have some bugs

#### 14 June 2013:
- Added POST /locations/ To Add A Location Thanks to Richard Vinke



index.php
<?php
use FAAPI\Inventory;
use FAAPI\InventoryLocations;
use FAAPI\Category;
use FAAPI\TaxTypes;
use FAAPI\TaxGroups;
use FAAPI\Customers;
use FAAPI\Suppliers;
use FAAPI\BankAccounts;
use FAAPI\GLAccounts;
use FAAPI\Currencies;
use FAAPI\InventoryCosts;
use FAAPI\Sales;
use FAAPI\Dimensions;
use FAAPI\Journal;
use FAAPI\ExchangeRates;
use FAAPI\GLQueries;

/**********************************************
Author: Andres Amaya
Name: SASYS REST API
Free software under GNU GPL
***********************************************/

/**
 * @SWG\Swagger(
 *   host="demo.saygoweb.com",
 *   basePath="/frontaccounting/modules/api",
 *   @SWG\Info(
 *     version="2.4-1.7",
 *     title="Front Accounting Simple API",
 *     description="This is a simple REST API as a Front Accounting module [https://github.com/andresamayadiaz/FrontAccountingSimpleAPI](https://github.com/andresamayadiaz/FrontAccountingSimpleAPI).",
 *     @SWG\License(
 *       name="GPL V2.0",
 *       url="https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html"
 *     )
 *   ),
 *   @SWG\ExternalDocumentation(
 *     description="Find out more about Front Accounting Simple API",
 *     url="https://github.com/andresamayadiaz/FrontAccountingSimpleAPI"
 *   )
 * )
 */

ini_set('html_errors', false);
ini_set('xdebug.show_exception_trace', 0);
// ini_set('xdebug.auto_trace', 2);

include_once('config_api.php');

global $security_areas, $security_groups, $security_headings, $path_to_root, $db, $db_connections;

$page_security = 'SA_API';

include_once(API_ROOT . "/session-custom.inc");
include_once(API_ROOT . "/vendor/autoload.php");

include_once(API_ROOT . "/util.php");

include_once(FA_ROOT . "/includes/date_functions.inc");
include_once(FA_ROOT . "/includes/data_checks.inc");

$rest = new \Slim\Slim(array(
    'log.enabled' => true,
    'mode' => 'debug',
    'debug' => true
));
$rest->setName('SASYS');

// API Login Hook
api_login();

$req = $rest->request();

define("RESULTS_PER_PAGE", 2);

class JsonToFormData extends \Slim\Middleware
{
    public function call()
    {
        $env = $this->app->environment();
        if (is_array($env['slim.input'])) {
            $env['slim.request.form_hash'] = $env['slim.input'];
        }
        $this->next->call();
    }
}

/*
The order of these 'add' calls is important, the JsonToFormData must be the
second Middleware called, which means it needs to be added first.
*/
$rest->add(new JsonToFormData());
$rest->add(new \Slim\Middleware\ContentTypes());

// API Routes
// ---------------------------------- Items -----------------------------------
$rest->container->singleton('inventory', function () {
    return new Inventory();
});
$rest->group('/inventory', function () use ($rest) {
    // Get Items
    $rest->get('/', function () use ($rest) {
        $rest->inventory->get($rest);
    });
    // Get Specific Item by Stock Id
    $rest->get('/:id', function ($id) use ($rest) {
        $rest->inventory->getById($rest, $id);
    });
    // Add Item
    $rest->post('/', function () use ($rest) {
        $rest->inventory->post($rest);
    });
    // Edit Specific Item
    $rest->put('/:id', function ($id) use ($rest) {
        $rest->inventory->put($rest, $id);
    });
    // Delete Specific Item
    $rest->delete('/:id', function ($id) use ($rest) {
        $rest->inventory->delete($rest, $id);
    });
});
// ---------------------------------- Items -----------------------------------

// ---------------------------- Inventory Locations ---------------------------
$rest->container->singleton('inventoryLocations', function () {
    return new InventoryLocations();
});
$rest->group('/locations', function () use ($rest) {
    // Get Locations
    $rest->get('/', function () use ($rest) {
        $rest->inventoryLocations->get($rest);
    });

    // Add Location, added by Richard Vinke
    $rest->post('/', function () use ($rest) {
        $rest->inventoryLocations->post($rest);
    });
});
// ---------------------------- Inventory Locations ---------------------------

// ----------------------------- Stock Adjustments ----------------------------
// Add Stock Adjustment
$rest->post('/stock/', function () use ($rest) {
    include_once(API_ROOT . "/inventory.inc");
    stock_adjustment_add();
});
// ----------------------------- Stock Adjustments ----------------------------

// ------------------------------ Item Categories -----------------------------
$rest->container->singleton('category', function () {
    return new Category();
});
$rest->group('/category', function () use ($rest) {
    // Get Items Categories
    $rest->get('/', function () use ($rest) {
        $rest->category->get($rest);
    });
    // Get Specific Item Category
    $rest->get('/:id', function ($id) use ($rest) {
        $rest->category->getById($rest, $id);
    });
    // Add Item Category
    $rest->post('/', function () use ($rest) {
        $rest->category->post($rest);
    });
    // Edit Item Category
    $rest->put('/:id', function ($id) use ($rest) {
        $rest->category->put($rest, $id);
    });
    // Delete Item Category
    $rest->delete('/:id', function ($id) use ($rest) {
        $rest->category->delete($rest, $id);
    });
});
// ------------------------------ Item Categories -----------------------------

// --------------------------------- Tax Types --------------------------------
// Tax Types
$rest->container->singleton('taxTypes', function () {
    return new TaxTypes();
});
$rest->group('/taxtypes', function () use ($rest) {
    // Get All Item Tax Types
    $rest->get('/', function () use ($rest) {
        $rest->taxTypes->get($rest);
    });
    // Get Specific Tax Type
    $rest->get('/:id', function ($id) use ($rest) {
        $rest->taxTypes->getById($rest, $id);
    });
});
// --------------------------------- Tax Types --------------------------------

// --------------------------------- Tax Groups -------------------------------
// Tax Groups
$rest->container->singleton('taxGroups', function () {
    return new TaxGroups();
});

// Get All Tax Groups
$rest->get('/taxgroups/', function () use ($rest) {
    $rest->taxGroups->get($rest);
});
// --------------------------------- Tax Groups -------------------------------

// --------------------------------- Customers --------------------------------
$rest->container->singleton('customers', function () {
    return new Customers();
});
$rest->group('/customers', function () use ($rest) {
    // Get Customer General Info
    $rest->get('/:id', function ($id) use ($rest) {
        $rest->customers->getById($rest, $id);
    });
    // All Customers
    $rest->get('/', function () use ($rest) {
        $rest->customers->get($rest);
    });
    // Add Customer
    $rest->post('/', function () use ($rest) {
        $rest->customers->post($rest);
    });
    // Edit Customer
    $rest->put('/:id', function ($id) use ($rest) {
        $rest->customers->put($rest, $id);
    });
    // Delete Customer
    $rest->delete('/:id', function ($id) use ($rest) {
        $rest->customers->delete($rest, $id);
    });
    // Get Customer Branches
    $rest->get('/:id/branches/', function ($id) use ($rest) {
        $rest->customers->getBranches($rest, $id);
    });
});
// --------------------------------- Customers --------------------------------

// --------------------------------- Suppliers --------------------------------
$rest->container->singleton('suppliers', function () {
    return new Suppliers();
});
$rest->group('/suppliers', function () use ($rest) {
    // All Suppliers
    $rest->get('/', function () use ($rest) {
        $rest->suppliers->get($rest);
    });
    // Get Supplier General Info
    $rest->get('/:id', function ($id) use ($rest) {
        $rest->suppliers->getById($rest, $id);
    });
    // Add Supplier
    $rest->post('/', function () use ($rest) {
        $rest->suppliers->post($rest);
    });
    // Edit Supplier
    $rest->put('/:id', function ($id) use ($rest) {
        $rest->suppliers->put($rest, $id);
    });
    // Delete Supplier
    $rest->delete('/:id', function ($id) use ($rest) {
        $rest->suppliers->delete($rest, $id);
    });
    // Get Supplier Contacts
    $rest->get('/:id/contacts/', function ($id) use ($rest) {
        $rest->suppliers->getContacts($rest, $id);
    });
});
// --------------------------------- Suppliers --------------------------------

// ------------------------------- Bank Accounts ------------------------------
$rest->container->singleton('bankAccounts', function () {
    return new BankAccounts();
});
$rest->group('/bankaccounts', function () use ($rest) {
    // Get All Bank Accounts
    $rest->get('/', function () use ($rest) {
        $rest->bankAccounts->get($rest);
    });
    // Get Specific Bank Account
    $rest->get('/:id', function ($id) use ($rest) {
        $rest->bankAccounts->getById($rest, $id);
    });
    // Insert Bank Account
    $rest->post('/', function () use ($rest) {
        $rest->bankAccounts->post($rest);
    });
    // Update Bank Account
    $rest->put('/:id', function ($id) use ($rest) {
        $rest->bankAccounts->put($rest, $id);
    });
    // Delete Bank Account
    $rest->delete('/:id', function ($id) use ($rest) {
        $rest->bankAccounts->delete($rest, $id);
    });
});
// ------------------------------- Bank Accounts ------------------------------

// -------------------------------- GL Accounts -------------------------------
$rest->container->singleton('glAccounts', function () {
    return new GLAccounts();
});
$rest->group('/glaccounts', function () use ($rest) {
    // Get All GL Accounts
    $rest->get('/', function () use ($rest) {
        $rest->glAccounts->get($rest);
    });
    // Get Specific GL Account
    $rest->get('/:id', function ($id) use ($rest) {
        $rest->glAccounts->getById($rest, $id);
    });
    // Insert GL Account
    $rest->post('/', function () use ($rest) {
        $rest->glAccounts->post($rest);
    });
    // Update GL Account
    $rest->put('/:id', function ($id) use ($rest) {
        $rest->glAccounts->put($rest, $id);
    });
    // Delete GL Account
    $rest->delete('/:id', function ($id) use ($rest) {
        $rest->glAccounts->delete($rest, $id);
    });
});
// Get GL Account Types
$rest->get('/glaccounttypes/', function () use ($rest) {
    $rest->glAccounts->getTypes($rest);
});
$rest->container->singleton('glQueries', function () {
    return new GLQueries();
});
$rest->group('/glquery', function () use ($rest) {
    // Get Trial Balance
    $rest->get('/trialbalance/:start/:end/', function ($start, $end) use ($rest) {
        $rest->glQueries->trialBalance($rest, $start, $end);
    });
});
// -------------------------------- GL Accounts -------------------------------

// -------------------------------- Currencies --------------------------------
$rest->container->singleton('currencies', function () {
    return new Currencies();
});
$rest->group('/currencies', function () use ($rest) {
    // Get All Currencies
    $rest->get('/', function () use ($rest) {
        $rest->currencies->get($rest);
    });
    // Get Specific Currency
    $rest->get('/:id', function ($id) use ($rest) {
        $rest->currencies->getById($rest, $id);
    });
});
// -------------------------------- Currencies --------------------------------
// ------------------------------ Exchange Rates ------------------------------
$rest->container->singleton('exchangerates', function () {
    return new ExchangeRates();
});
$rest->group('/exchangerates', function () use ($rest) {
    // Get Current ExchangeRate
    $rest->get('/:currency/current', function ($currency) use ($rest) {
        $rest->exchangerates->getCurrent($rest, $currency);
    });
    // Get ExchangeRate
    $rest->get('/:currency/:id', function ($currency, $id) use ($rest) {
        $rest->exchangerates->getById($rest, $currency, $id);
    });
    // Insert ExchangeRate
    $rest->post('/:currency/', function ($currency) use ($rest) {
        $rest->exchangerates->post($rest, $currency);
    });
    // Edit ExchangeRate
    $rest->put('/:currency/:id', function ($currency, $id) use ($rest) {
        $rest->exchangerates->put($rest, $currency, $id);
    });
    // Delete ExchangeRate
    $rest->delete('/:currency/:id', function ($currency, $id) use ($rest) {
        $rest->exchangerates->delete($rest, $currency, $id);
    });
    // All exchangerates for the given currency
    $rest->get('/:currency/', function ($currency) use ($rest) {
        $rest->exchangerates->getAllByCurrency($rest, $currency);
    });
});
// Get Last Exchange Rate (Legacy route)
$rest->get('/exrates/:curr_abrev', function ($curr_abrev) use ($rest) {
    $rest->currencies->getLastExchangeRate($rest, $curr_abrev);
});
// ------------------------------ Exchange Rates ------------------------------

// ----- ------------------------ Inventory Costs -----------------------------
$rest->container->singleton('inventoryCosts', function () {
    return new InventoryCosts();
});
$rest->group('/itemcosts', function () use ($rest) {
    // Get Item Cost
    $rest->get('/:id', function ($id) use ($rest) {
        $rest->inventoryCosts->getById($rest, $id);
    });
    // Update Item Cost
    $rest->put('/:id', function ($id) use ($rest) {
        $rest->inventoryCosts->put($rest, $id);
    });
});
// ----- ------------------------ Inventory Costs -----------------------------

// ------------------------------- Assets -------------------------------
// Fixed Assets
function assets_supported()
{
    global $path_to_root;
    return file_exists($path_to_root . '/modules/asset_register');
}
if (assets_supported()) {
    // Get Fixed Asset
    $rest->get('/assets/:id', function ($id) use ($rest) {
        include_once(API_ROOT . "/assets.inc");
        assets_get($id);
    });
    // Insert Fixed Asset
    $rest->post('/assets/', function () use ($rest) {
        include_once(API_ROOT . "/assets.inc");
        assets_add();
    });
    // Get Asset Types
    $rest->get('/assettypes/', function () use ($rest) {
        global $req;
        include_once(API_ROOT . "/assets.inc");

        $page = $req->get("page");

        if ($page == null) {
            assettypes_all();
        } else {
            // If page = 1 the value will be 0, if page = 2 the value will be 1, ...
            $from = -- $page * RESULTS_PER_PAGE;
            assettypes_all($from);
        }
    });
}
// ------------------------------- Assets -------------------------------

// ------------------------------- Sales --------------------------------
$rest->container->singleton('sales', function () {
    return new Sales();
});
$rest->group('/sales', function () use ($rest) {
    // Get Sales Header and Details
    $rest->get('/:trans_no/:trans_type', function ($trans_no, $trans_type) use ($rest) {
        $rest->sales->getById($rest, $trans_no, $trans_type);
    });
    // Insert Sales
    $rest->post('/', function () use ($rest) {
        $rest->sales->post($rest);
    });
    // Edit Sales
    $rest->put('/:trans_no/:trans_type', function ($trans_no, $trans_type) use ($rest) {
        $rest->sales->put($rest, $trans_no, $trans_type);
    });
    // Cancel Sales
    $rest->delete('/:branch_id/:uuid', function ($branch_id, $uuid) use ($rest) {
        $rest->sales->delete($rest, $branch_id, $uuid);
    });
    // All Sales
    $rest->get('/:trans_type/', function ($trans_type) use ($rest) {
        $rest->sales->get($rest, $trans_type);
    });
});
// ------------------------------- Sales --------------------------------

// ----------------------------- Dimensions -----------------------------
$rest->container->singleton('dimensions', function () {
    return new Dimensions();
});
$rest->group('/dimensions', function () use ($rest) {
    // Get Dimension
    $rest->get('/:ref', function ($ref) use ($rest) {
        $rest->dimensions->getById($rest, $ref);
    });
    // Insert Dimension
    $rest->post('/', function () use ($rest) {
        $rest->dimensions->post($rest);
    });
    // Edit Dimension
    $rest->put('/:ref', function ($ref) use ($rest) {
        $rest->dimensions->put($rest, $ref);
    });
    // Delete Dimension
    $rest->delete('/:ref', function ($ref) use ($rest) {
        $rest->dimensions->delete($rest, $ref);
    });
    // All Dimensions
    $rest->get('/', function () use ($rest) {
        $rest->dimensions->get($rest);
    });
});
// ----------------------------- Dimensions -----------------------------

// ------------------------------ Journal -------------------------------
$rest->container->singleton('journal', function () {
    return new Journal();
});
$rest->group('/journal', function () use ($rest) {
    // Get All Journal Entries
    $rest->get('/', function () use ($rest) {
        $rest->journal->get($rest);
    });
    // Get Specific Journal Entry
    $rest->get('/:type/:id', function ($type, $id) use ($rest) {
        $rest->journal->getById($rest, $type, $id);
    });
    // Insert Journal Entry
    $rest->post('/', function () use ($rest) {
        $rest->journal->post($rest);
    });
    // Update Journal Entry
    $rest->put('/:id', function ($id) use ($rest) {
        $rest->journal->put($rest, $id);
    });
    // Delete Journal Entry
    $rest->delete('/:type/:id', function ($type, $id) use ($rest) {
        $rest->journal->delete($rest, $type, $id);
    });
});
// ------------------------------ Journal -------------------------------

// Init API
$rest->run();
